packages/logger/src/logger/ogma.ts

```diff
@@ -154,15 +164,29 @@ export class Ogma {
     if (typeof message === 'object' && !(message instanceof Error)) {
       message = '\n' + JSON.stringify(message, this.circularReplacer(), 2);
     }
-    application = this.toStreamColor(application || this.options.application, Color.YELLOW);
     context = this.toStreamColor(context || this.options.context, Color.CYAN);
     correlationId &&= this.wrapInBrackets(correlationId);
 
-    const hostname = this.toStreamColor(this.hostname, Color.MAGENTA);
     const timestamp = this.wrapInBrackets(this.getTimestamp());
-    return `${timestamp} ${formattedLevel} ${hostname} ${application} ${this.wrapInBrackets(
-      this.pid.toString(),
-    )} ${correlationId} ${context} ${message}`;
+
+    let outputTuple = [timestamp, formattedLevel];
+
+    if (this.options.logHostname) {
+      outputTuple.push(this.toStreamColor(this.hostname, Color.MAGENTA));
+    }
+
+    if (this.options.logApplication) {
+      outputTuple.push(this.toStreamColor(application || this.options.application, Color.YELLOW));
+    }
+
+    if (this.options.logPid) {
+      outputTuple.push(this.wrapInBrackets(this.pid.toString()));
+    }
+
+    outputTuple = outputTuple.concat([correlationId, context, message]);
+
+    // [timestamp, formattedLevel, hostname?, application?, pid?, correlationId, context, message]
+    return outputTuple.join(' ');
```

Comment:
I think my only concern here is how is performance impacted? I can run the benchmark on my machine and see what kind of difference we get

Reply:
sounds good

the other options was to add whitespace to every element and concat result string almost as it was concatenated before, i.e. 
`${timestamp}${formattedLevel}${hostname} ...`
in case logVal=false -> Val = '', so we won't have excessive whitespaces

Reply:
Okay, got around to running the benchmarks again. This adds a decent bit of overhead, can you think of any way to make this faster? I believe `Array.prototype.join` is the slowdown here. Would it be faster to use a `for` loop instead? 

<details>

<summary>Benchmarks</summary>

<h2>Previous Commit</h2>

```
| Logger    | Simple  | Long     | JSON    | Deep     |
| --------- | ------- | -------- | ------- | -------- |
| Bunyan    | 383.3ms | 4037.3ms | 411.5ms | 2000.3ms |
| Ogma      | 177.0ms | 1790.2ms | 338.0ms | 467.4ms  |
| OgmaMasks | 171.2ms | 1846.6ms | 315.1ms | 695.4ms  |
| Pino      | 126.6ms | 4056.8ms | 470.4ms | 1385.1ms |
| Winston   | 398.4ms | 4724.1ms | 337.3ms | 462.8ms  |
```

<h2>Current Code</h2>

```
| Logger    | Simple  | Long     | JSON    | Deep     |
| --------- | ------- | -------- | ------- | -------- |
| Bunyan    | 368.8ms | 5243.0ms | 414.7ms | 2539.6ms |
| Ogma      | 365.5ms | 2979.6ms | 500.1ms | 669.1ms  |
| OgmaMasks | 321.7ms | 2761.6ms | 512.1ms | 713.4ms  |
| Pino      | 113.4ms | 4789.3ms | 120.1ms | 2067.9ms |
| Winston   | 545.8ms | 7506.3ms | 278.2ms | 708.9ms  |
```

I do see that Winston has pretty vaariable times too, but pino and bunyan seem pretty close to their old timings, so I'd think this probably isn't an issue with the system running the bench. I should really look into putting these into CI somehow

</details>

Reply:
Hmm, a preliminary test with a `for` loop didn't prove to improve the speed that much. I wonder if there's a different way we can approach this. Maybe a general `customFormat` method that gets called only if it exists, and only for the string output mode (letting JSON keep all the extra metadata as it's mainly for machine consumption anyways?)

Reply:
I think if we can get away from the multiple `if` statements it would help with the speed. And we can just pass all of the properties we already have to this `customFormat` method. It's worth noting the `customFormat` method should then be in charge of bracketing/colorizing the string if that's what the dev would want, rather than have us pass in a formatted value already, I think

Reply:
kk, will check out how to make it faster.
How do u perform benchmarking?
I've tried `nx run benchmark-logger:bench`, `pnpm benchmark` - getting errors in both bases (different errors though).

Reply:
Yeah, I realized the nx target it missing the `-r module-alias/register` necessary to make it work.
```json
"benchmark-logger": {
      "root": "benchmarks/logger",
      "sourceRoot": "benchmarks/logger/src",
      "type": "library",
      "targets": {
        "bench": {
          "executor": "@nrwl/workspace:run-commands",
          "options": {
            "cwd": "benchmarks/logger",
            "commands": [
              "../../node_modules/.bin/tsc -p tsconfig.json",
              "node -r module-alias/register dist/benchmarks/logger/src/index"
            ],
            "parallel": false
          }
        }
      },
      "implicitDependencies": []
    },
```

Reply:
Done some tests, bulk part of slowdown is caused by array.join
IF statements have very minor contribution.
Done some tests:

```
template string: 730.325ms
array join: 1.222s
```

How about rewriting formatStream to get rid of array.join and as for JSON format - don't see much of a slowdown. Ther results are actually fluctuating and sometime I'm getting better time with IFs than 'main'.

Reply:
That sounds like a good approach to me. Shaving off half a second feels like the right direction here